### PR TITLE
document new features in `cupy.fft` and `cupyx.scipy.fftpack`

### LINKED ISSUE #1988: The `overwrite_x` argument of FFT functions is deceptive in certain conditions
In the current master branch, several functions in `cupy.scipy.fftpack` support the argument `overwrite_x` as requested by me earlier (#1669). Ideally, one would expect that after executing the following line
```python
cupy.scipy.fftpack.fft(a, overwrite_x=True)
```
the input array `a` is correctly transformed **in place** (meaning `a=cupy.fft.fft(a)` but without extra allocation).

However, this expectation is not always true. Consider this example when `a` has more than 1 axis and we are transforming along an axis not the last one:
```python
a = cupy.random.random((10, 10)).astype(cupy.complex128)
cupy.scipy.fftpack.fft(a, axis=0, overwrite_x=True)
```
In this case, internally we do a swap followed by a copy (i.e. allocate a new array):
https://github.com/cupy/cupy/blob/f4acb4e86c64bf0239f8f25418bc301d63a92ead/cupy/fft/fft.py#L76-L80
So in the end with `overwrite_x=True` we return the reference to the new array, not the input array. This means that after the operation above `a` does not really hold the correctly transformed result (i.e. the result is not the same as `a=cupy.fft.fft(a, axis=0)` or its NumPy equivalent) because it's the copied array that got transformed.

I can think of two workarounds now. The first one is simpler but counter-intuitive: we just advocate the following usage
```python
a = cupy.scipy.fftpack.fft(a, axis=whatever, overwrite_x=True)
```
which will always be correct. The second one is to swap the pointers after copy:
```python 
if a.base is not None or not a.flags.c_contiguous: 
    b = a.copy() 
    if overwrite_x:
        temp = a.data
        a.data = b.data
        b.data = temp
    else:
        a = b
```
so that the expected usage can work. However, neither workarounds can truly avoid a copy, which could cause performance penalty and require extra memory merely for accommodating the intermediate step. 

If I remember it correctly, all FFT functions (1d, 2d, 3d) are affected, not just `cupy.scipy.fftpack.fft`, but I haven't looked into the call paths for other cases carefully.

### PR DESCRIPTION
**Pull request by leofang**
addresses #1669 and closes #1988.

### CODE REVIEW

**Comment by leofang** [comment]
related PR: #1746, #1942, #2033

**Comment by leofang** [comment]
attn: @grlee77

File: docs/source/reference/fftpack.rst [hunk 1]
```diff
@@ -24,4 +24,8 @@ Fast Fourier Transforms
 
 Code compatibility features
 ---------------------------
-The ``get_fft_plan`` function has no counterpart in ``scipy.fftpack``. It returns a cuFFT plan that can be passed to the FFT functions in this module (using the argument ``plan``) to accelarate the computation.
+1. The ``get_fft_plan`` function has no counterpart in ``scipy.fftpack``. It returns a cuFFT plan that can be passed to the FFT functions in this module (using the argument ``plan``) to accelarate the computation.
+
+2. The boolean switch ``cupy.fft.config.enable_nd_planning`` also affects the FFT functions in this module, see :doc:`./fft`. This switch is neglected when planning "manually" using ``get_fft_plan``.
+
+3. Like in ``scipy.fftpack``, all FFT functions in this module have an optional argument ``overwrite_x`` (default is ``False``), which has the same semantics as in ``scipy.fftpack``: when it is set to ``True``, the input array ``x`` *can* (not *will*) be destroyed and replaced by the output. Therefore, to guarantee an in-place FFT is successfully performed, one should always re-assign the input: ``x = cupyx.scipy.fftpack.fft(x, ..., overwrite_x=True, ...)``.
```
Comment (grlee77) [inline docs/source/reference/fftpack.rst]: One nitpick about the last sentence: It is not really an "in-place FFT" if it is necessary to reassign the input. 

Maybe replace the last sentence with:

For this reason, when an in-place FFT is desired, the user should always reassign the input in the following manner: ``x = cupyx.scipy.fftpack.fft(x, ..., overwrite_x=True, ...)``.

File: docs/source/reference/fft.rst [hunk 2]
```diff
@@ -69,3 +69,7 @@ Code compatibility features
 ---------------------------
 FFT functions of NumPy alway return numpy.ndarray which type is ``numpy.complex128`` or ``numpy.float64``.
 CuPy functions do not follow the behavior, they will return ``numpy.complex64`` or ``numpy.float32`` if the type of the input is ``numpy.float16``, ``numpy.float32``, or ``numpy.complex64``.
+
+In addition, when transforming over more than 1 axis ``cupy.fft`` will attempt to generate a *cuFFT plan* internally (see the `cuFFT documentation`_ for detail) to accelarate the computation. This is enabled by default but can be turned off by setting ``cupy.fft.config.enable_nd_planning = False``. This feature is a deviation from NumPy which has no planning.
```
[the remark this answers is not in the document]
Comment (leofang) [reply docs/source/reference/fft.rst]: > Technically the 1D cases also use cuFFT plans.

Right, don't know why I overlooked it...Thanks!

File: docs/source/reference/fftpack.rst (hunk 1 above)
Comment (leofang) [reply docs/source/reference/fftpack.rst]: > One nitpick about the last sentence: It is not really an "in-place FFT" if it is necessary to reassign the input.

Yeah I know...This was precisely I raised #1988...

**Comment by leofang** [comment]
Thanks a lot @grlee77!

**Comment by leofang** [comment]
I think it's better to highlight in the doc that the `plan` argument is currently experimental. (See https://github.com/cupy/cupy/pull/2033#issuecomment-467367627)

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM! Thanks a lot!

**Pull request merged**